src/main/java/org/folio/fqm/service/EntityTypeFlatteningService.java

```diff
@@ -0,0 +1,242 @@
+      .orElseThrow(() -> new EntityTypeNotFoundException(entityTypeId));
+    EntityType flattenedEntityType = new EntityType()
+      .id(originalEntityType.getId())
+      .name(originalEntityType.getName())
+      ._private(originalEntityType.getPrivate())
+      .defaultSort(originalEntityType.getDefaultSort())
+      .columns(originalEntityType.getColumns()) // TODO: probably remove
+      .idView(originalEntityType.getIdView())
+      .customFieldEntityTypeId(originalEntityType.getCustomFieldEntityTypeId())
+      .labelAlias(originalEntityType.getLabelAlias())
+      .root(originalEntityType.getRoot())
+      .sourceView(originalEntityType.getSourceView()) // Possibly unneeded
+      .sourceViewExtractor(originalEntityType.getSourceViewExtractor()); // Possibly unneeded
+
+    List<EntityTypeColumn> finalColumns = new ArrayList<>();
+    for (EntityTypeSource source : originalEntityType.getSources()) {
+      if (source.getType().equals("db")) {
+        Pair<EntityTypeSource, List<EntityTypeColumn>> updatePair = getConvertedSourceAndColumns(originalEntityType, source, null, false); // TODO: think about this, may not be able to hardcode false here
+        flattenedEntityType.addSourcesItem(updatePair.component1());
+        finalColumns.addAll(updatePair.component2());
+      } else {
+        UUID sourceEntityTypeId = UUID.fromString(source.getId());
+        EntityType flattenedSourceDefinition = getFlattenedEntityType(sourceEntityTypeId, false)
+          .orElseThrow(() -> new EntityTypeNotFoundException(sourceEntityTypeId));
+
+        // If an entity type source contains multiple db sources, then we need to keep the original alias in order to
+        // distinguish the different targets. Frequently, it will likely only have one db source. In this case we
+        // can use the outer alias only, in order to keep field names more concise
+        boolean keepOriginalAlias = countDbSources(flattenedSourceDefinition) > 1;
+
+        for (EntityTypeSource subSource : flattenedSourceDefinition.getSources()) {
+          String oldAlias = subSource.getAlias();
+          Pair<EntityTypeSource, List<EntityTypeColumn>> updatePair = getConvertedSourceAndColumns(flattenedSourceDefinition, subSource, source, keepOriginalAlias);
+          String newAlias = updatePair.component1().getAlias();
+          if (!oldAlias.equals(newAlias)) {
+            updateOtherSources(oldAlias, newAlias, flattenedSourceDefinition.getSources());
+          }
+          flattenedEntityType.addSourcesItem(updatePair.component1());
+          finalColumns.addAll(updatePair.component2());
+        }
+      }
+    }
+
+    flattenedEntityType.columns(finalColumns);
+    if (doFinalRenames) {
+      List<EntityTypeColumn> convertedColumns = finalColumnConversion(flattenedEntityType);
+      flattenedEntityType.columns(convertedColumns);
+    }
+    return Optional.of(flattenedEntityType);
+  }
+
+  public String getJoinClause(EntityType flattenedEntityType) {
```

Comment:
Nit: I don't like this name, since afaik, there's no such thing as a join clause